src/blob/index/bptree/node.rs

```diff
@@ -0,0 +1,117 @@
+use super::prelude::*;
+
+#[derive(Debug, PartialEq)]
+pub(super) struct Node {
+    keys: Vec<Vec<u8>>,
+    offsets: Vec<u64>,
+}
+
+impl Node {
+    pub(super) fn new(keys: Vec<Vec<u8>>, offsets: Vec<u64>) -> Self {
+        assert!(keys.len() != 0);
+        assert_eq!(keys.len() + 1, offsets.len());
+        Self { keys, offsets }
+    }
+
+    pub(super) fn key_offset(&self, key: &[u8]) -> u64 {
+        let mut left = 0i32;
+        let mut right = self.keys.len() as i32 - 1;
+        while left <= right {
+            let mid = (left + right) / 2;
+            match key.cmp(&self.keys[mid as usize]) {
+                CmpOrdering::Equal => return self.offsets[mid as usize + 1],
+                CmpOrdering::Greater => left = mid + 1,
+                CmpOrdering::Less => right = mid - 1,
+            }
+        }
+        if left == 0 || key.cmp(&self.keys[left as usize - 1]) != CmpOrdering::Less {
+            self.offsets[left as usize] as u64
+        } else {
+            self.offsets[(left + 1) as usize] as u64
+        }
+    }
+
+    pub(super) fn serialize(&self) -> Result<Vec<u8>> {
+        let mut buf = Vec::new();
+        let meta = NodeMeta::new(self.keys.len() as u64);
+        buf.extend_from_slice(&serialize(&meta)?);
+        self.keys.iter().fold(&mut buf, |acc, key| {
```

Comment:
Why not flat_map?

Reply:
I am not sure, that I understood you right, but if I'll iterate through keys with flat_map (and get something like `key_byte` sequence) I'll need to collect there somewhere, because I can't extend from iterator (and this collection step will give one more memory allocation)

Reply:
You can extend from iterator by `extend` method, why not?

Reply:
Forgot about this method, did it!